### ISSUE TITLE
Dramatic bash prompt performance regression in 2.26-2: ~four-five seconds wait after each command

### ISSUE THREAD

**Issue by jakub-g** [git-for-windows/git#2595]
This is similar to a few issues reported in the past, but since it's a regression in a new release, I'm opening a new bug.

After upgrading from 2.26-1 to 2.26-2, evaluating bash prompt after the command executed takes 5 seconds. When I downgrade, the problem goes away.

I followed https://github.com/msysgit/msysgit/wiki/Diagnosing-why-Git-is-so-slow#trace-executions-in-the-bash-startup

The two commands below take about 2.5s each, regardless of folder I am in:

```
git rev-parse --is-inside-work-tree
git rev-parse --abbrev-ref HEAD
```

# Setup

```
Git-2.26.0-64-bit.exe good
Git-2.26.1-64-bit.exe good
Git-2.26.2-64-bit.exe bad
```

```
$ git --version --build-options

git version 2.26.2.windows.1
cpu: x86_64
built from commit: 4267c88d0d8936ae723bba55e0576b1c4f3d4e1a
sizeof-long: 4
sizeof-size_t: 8
```

 - Which version of Windows: Win10 64 bit

```
$ cmd.exe /c ver
Microsoft Windows [Version 10.0.18363.778]
```

 - Options

```
$ cat /etc/install-options.txt
Editor Option: VIM
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: LFOnly
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

### Details

 - Which terminal/shell are you running Git from?

Bash or ConEmu, same behavior

 - What commands did you run to trigger this issue?

any command

**Comment by jakub-g** [comment]
According to `time` it's more like 1.8s each:

```
$ time GIT_TRACE=1 git rev-parse --is-inside-work-tree                                                              
+ GIT_TRACE=1                                                                                                       
+ git rev-parse --is-inside-work-tree                                                                               
15:20:27.092127 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/bin            
15:20:27.108474 git.c:439               trace: built-in: git rev-parse --is-inside-work-tree                        
true                                                                                                                
                                                                                                                    
real    0m1.864s                                                                                                    
user    0m0.000s                                                                                                    
sys     0m0.031s                                                                                                    

$ time GIT_TRACE=1 git rev-parse --abbrev-ref HEAD
+ GIT_TRACE=1
+ git rev-parse --abbrev-ref HEAD
15:21:37.769957 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/bin
15:21:37.785754 git.c:439               trace: built-in: git rev-parse --abbrev-ref HEAD
<branchname>

real    0m1.811s
user    0m0.000s
sys     0m0.047s
```

**Comment by dscho** [comment]
Would you mind re-running with [`GIT_TRACE_PERFORMANCE`](https://git-scm.com/docs/git#Documentation/git.txt-codeGITTRACEPERFORMANCEcode)?

**Comment by jakub-g** [comment]
Ok, gonna do that tomorrow

**Comment by jakub-g** [comment]
Alright so I copied 2.26.1 to `/c/Program Files/Git2261` and installed 2.26.2 to `/c/Program Files/Git`

- perf numbers from git trace are similar, so the problem is outside of it
- there's about 1s wait *before any git output* and 1s wait *after git output*
- `PageFaultCount` is up 18x in new version (from ~7k to ~130k)
- strange thing is that executing `git` i have 2s slowdown but executing ` '/c/Program Files/Git/bin/git'` I have 4s slowdown

```
$ time GIT_TRACE2_PERF=1 '/c/Program Files/Git2261/bin/git' rev-parse --is-inside-work-tree
+ GIT_TRACE2_PERF=1
+ '/c/Program Files/Git2261/bin/git' rev-parse --is-inside-work-tree
19:19:11.979289 common-main.c:48             | d0 | main                     | version      |     |           |           |              | 2.26.1.windows.1
19:19:11.980290 common-main.c:49             | d0 | main                     | start        |     |  0.009829 |           |              | git.exe rev-parse --is-inside-work-tree
19:19:12.031285 ..._win32_process_info.c:118 | d0 | main                     | data_json    | r0  |  0.061264 |  0.061264 | process      | windows/ancestry:["git.exe","bash.exe","bash.exe","bash.exe","ConEmuC64.exe","ConEmu64.exe","explorer.exe"]
19:19:12.032286 git.c:440                    | d0 | main                     | cmd_name     |     |           |           |              | rev-parse (rev-parse)
19:19:12.033289 repository.c:130             | d0 | main                     | def_repo     | r1  |           |           |              | worktree:C:/git/player
19:19:12.035327 ..._win32_process_info.c:166 | d0 | main                     | data_json    | r1  |  0.064688 |  0.064688 | process      | windows/memory:{"PageFaultCount":7166,"PeakWorkingSetSize":7786496,"PeakPagefileUsage":4239360}
19:19:12.035327 git.c:674                    | d0 | main                     | exit         |     |  0.064730 |           |              | code:0
19:19:12.035327 trace2/tr2_tgt_perf.c:213    | d0 | main                     | atexit       |     |  0.064742 |           |              | code:0
true

real    0m0.180s
user    0m0.015s
sys     0m0.047s

$ time GIT_TRACE2_PERF=1 git rev-parse --is-inside-work-tree
+ GIT_TRACE2_PERF=1
+ git rev-parse --is-inside-work-tree
19:19:20.845162 common-main.c:48             | d0 | main                     | version      |     |           |           |              | 2.26.2.windows.1
19:19:20.846163 common-main.c:49             | d0 | main                     | start        |     |  0.016231 |           |              | 'C:\Program Files\Git\mingw64\bin\git.exe' rev-parse --is-inside-work-tree
19:19:20.896163 ..._win32_process_info.c:118 | d0 | main                     | data_json    | r0  |  0.065803 |  0.065803 | process      | windows/ancestry:["bash.exe","bash.exe","bash.exe","ConEmuC64.exe","ConEmu64.exe","explorer.exe"]
19:19:20.896163 git.c:440                    | d0 | main                     | cmd_name     |     |           |           |              | rev-parse (rev-parse)
19:19:20.898162 repository.c:130             | d0 | main                     | def_repo     | r1  |           |           |              | worktree:C:/git/player
19:19:20.900163 ..._win32_process_info.c:166 | d0 | main                     | data_json    | r1  |  0.070221 |  0.070221 | process      | windows/memory:{"PageFaultCount":129175,"PeakWorkingSetSize":8126464,"PeakPagefileUsage":4775936}
19:19:20.900163 git.c:674                    | d0 | main                     | exit         |     |  0.070315 |           |              | code:0
19:19:20.900163 trace2/tr2_tgt_perf.c:213    | d0 | main                     | atexit       |     |  0.070341 |           |              | code:0
true

real    0m2.369s
user    0m0.015s
sys     0m0.077s

$ time GIT_TRACE2_PERF=1 '/c/Program Files/Git/bin/git' rev-parse --is-inside-work-tree
+ GIT_TRACE2_PERF=1
+ '/c/Program Files/Git/bin/git' rev-parse --is-inside-work-tree
19:21:40.351891 common-main.c:48             | d0 | main                     | version      |     |           |           |              | 2.26.2.windows.1
19:21:40.352892 common-main.c:49             | d0 | main                     | start        |     |  0.013496 |           |              | git.exe rev-parse --is-inside-work-tree
19:21:40.404854 ..._win32_process_info.c:118 | d0 | main                     | data_json    | r0  |  0.065111 |  0.065111 | process      | windows/ancestry:["git.exe","bash.exe","bash.exe","bash.exe","ConEmuC64.exe","ConEmu64.exe","explorer.exe"]
19:21:40.404854 git.c:440                    | d0 | main                     | cmd_name     |     |           |           |              | rev-parse (rev-parse)
19:21:40.405856 repository.c:130             | d0 | main                     | def_repo     | r1  |           |           |              | worktree:C:/git/player
19:21:40.407856 ..._win32_process_info.c:166 | d0 | main                     | data_json    | r1  |  0.068029 |  0.068029 | process      | windows/memory:{"PageFaultCount":128359,"PeakWorkingSetSize":8110080,"PeakPagefileUsage":4771840}
19:21:40.407856 git.c:674                    | d0 | main                     | exit         |     |  0.068059 |           |              | code:0
19:21:40.407856 trace2/tr2_tgt_perf.c:213    | d0 | main                     | atexit       |     |  0.068072 |           |              | code:0
true

real    0m4.432s
user    0m0.015s
sys     0m0.046s
```

**Comment by dscho** [comment]
>     * there's about 1s wait _before any git output_ and 1s wait _after git output_

That sounds awfully like one of those antimalware protection things going off in all the wrong ways.

Additionally, I do see two snapshots between v2.26.1 and v2.26.2 in https://wingit.blob.core.windows.net/files/index.html, would you mind testing those? (You can use the portable Git, that should make testing side-by-side easier.)

**Comment by jakub-g** [comment]
I just re-checked today and the problem is mysteriously gone. It could be indeed that it was the corporate antivirus blocking things, and malware definitions were updated in the meantime 🤔

Anyway, thx for having a look!

**Comment by dscho** [comment]
Is there a chance that you work on a domain-joined machine and that there were network issues?

**Comment by jakub-g** [comment]
The machine is domain-joined but I don't think there were any network issues.
(the issue would repo both with corporate VPN enabled and disabled)

**Comment by dscho** [comment]
I still guess that the slowness had something to do with the fact that the machine is domain-joined. I experienced similar issues in the past, but they were never pressing enough to go into full debug mode: just as in your case, the issues went away relatively quickly.

Having said that, I would love to have an MSYS2 runtime that does not try to reach the domain controller by default. It's probably a big reason why running shell scripts in Git for Windows is _such_ a drag: it just slows everything down.

**Comment by mfriedrich74** [comment]
Hi,

Git is slow if the global gitconfig and bashrc files are on a network drive
(check HOME variable. I usually set those files as offline available.
You may check if that is the case for you. Having HOME on a network drive
is rather common in a domain environment.

Best regards, Mike

**Comment by jakub-g** [comment]
To follow up on that, for info:

The same happened when I installed git 2.28.0. Immediate dramatic slowdown. So I uninstalled it (reverted to 2.26.x). Few days later, I installed 2.28.0 again, and it works normally.

I think my antivirus (F-Secure) is doing some kind of quarantine of the newly installed git binary and slowing it down for N hours until it decides it's "safe". Unfortunately I can't disable the AV (it's a corporate machine). I guess I need to update git this way each time: install, run, uninstall, wait a few days, install again 🤷‍♂️
